src/pkgcheck/checks/eclass.py

```diff
@@ -498,3 +500,85 @@ def feed(self, eclass):
         )
         if vars_missing_docs:
             yield EclassDocMissingVar(sorted(vars_missing_docs), eclass=eclass)
+
+
+class GoMissingDeps(results.VersionResult, results.Warning):
+    """Package sets ``GO_OPTIONAL`` but does not depend on ``dev-lang/go``."""
+
+    desc = "sets GO_OPTIONAL but does not depend on dev-lang/go"
+
+
+class RubyMissingDeps(results.VersionResult, results.Warning):
+    """Package sets ``RUBY_OPTIONAL`` but does not depend on ``dev-lang/ruby``
+    or ``virtual/rubygems``."""
+
+    desc = "sets RUBY_OPTIONAL but does not depend on dev-lang/ruby or virtual/rubygems"
+
+
+class RustMissingDeps(results.VersionResult, results.Warning):
+    """Package sets ``CARGO_OPTIONAL`` but does not depend on ``virtual/rust``."""
+
+    desc = "sets CARGO_OPTIONAL but does not depend on virtual/rust"
+
+
+class TmpfilesMissingDeps(results.VersionResult, results.Warning):
+    """Package sets ``TMPFILES_OPTIONAL`` but does not depend on ``virtual/tmpfiles``."""
+
+    desc = "sets TMPFILES_OPTIONAL but does not depend on virtual/tmpfiles"
+
+
+class EclassManualDepsCheck(Check):
+    """Check for missing deps when inheriting eclasses in special mode."""
+
+    _source = sources.EbuildParseRepoSource
+    known_results = frozenset(
+        {
+            GoMissingDeps,
+            RustMissingDeps,
+            RubyMissingDeps,
+            TmpfilesMissingDeps,
+        }
+    )
+
+    dependencies = (
+        # eclass, variable, one of deps, class
+        ("cargo", "CARGO_OPTIONAL", {"virtual/rust"}, RustMissingDeps),
+        ("go-module", "GO_OPTIONAL", {"dev-lang/go"}, GoMissingDeps),
+        ("ruby-ng", "RUBY_OPTIONAL", {"dev-lang/ruby", "virtual/rubygems"}, RubyMissingDeps),
+        ("tmpfiles", "TMPFILES_OPTIONAL", {"virtual/tmpfiles"}, TmpfilesMissingDeps),
+    )
+
+    def __init__(self, options, **kwargs):
+        super().__init__(options, **kwargs)
+
+        self.queries_by_eclass = defaultdict(list)
+        for eclass, variable, deps, cls in self.dependencies:
+            self.queries_by_eclass[eclass].append(
+                (
+                    bash.query(
+                        # has variable assignment to a variable named
+                        f'(variable_assignment name: (variable_name) @name (.eq? @name "{variable}"))'
```

Comment:
I really need to get up to speed on TS queries. I bet we can simplify a lot of checks with this1

Reply:
The best [documentation](https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries) I found.

And to perform fast tests, I use the [playground](https://tree-sitter.github.io/tree-sitter/playground).